src/UUPSProxy.sol

```diff
@@ -35,14 +38,26 @@ contract UUPSProxy is Proxy, IUUPSProxy {
      * - If `data` is empty, `msg.value` must be zero.
      */
     function initialize(address implementation, bytes memory data) public payable {
-        require(implementation != address(0), "New implementation cannot be the zero address");
-        require(_implementation() == address(0), "Already initialized");
+        if (implementation == address(0)) revert ImplementationCannotBeZeroAddress();
+        if (_implementation() != address(0)) revert AlreadyInitialized();
 
         ERC1967Utils.upgradeToAndCall(implementation, data);
     }
 
-    function getVerifiableProxySalt() public view returns (bytes32) {
-        return _SALT_SLOT.erc7201Slot().getBytes32Slot().value;
+    function getVerifiableProxyData() public view returns (bytes32 salt, address implementation) {
+        return (_SALT_SLOT.erc7201Slot().getBytes32Slot().value, _implementation());
+    }
+
+    function upgradeToAndCall(address newImplementation, bytes memory /*data*/ ) public payable {
+        if (newImplementation == address(0)) revert ImplementationCannotBeZeroAddress();
+        if (_implementation() == address(0)) revert ImplementationNotSet();
+
+        IProxyAuthorization newImpl = IProxyAuthorization(newImplementation);
+
+        if (!newImpl.canUpgradeFrom(_implementation())) revert InvalidUpgradeTargetForCurrentImplementation();
+
+        // forward the call to the implementation
+        _delegate(_implementation(), false);
```

Comment:
An alternative here would be to have an `upgradeFrom` method that the new implementation is expected to have, which is passed the address of the previous implementation (or, it can just get it from the storage slot); it can revert if it doesn't like the previous implementation or anything else about the upgrade. Thoughts?